spread-shellcheck

```diff
@@ -133,15 +140,74 @@ def checksection(data, env: Dict[str, str]):
                             stdout=subprocess.PIPE,
                             stdin=subprocess.PIPE,
                             shell=True)
-    stdout, _ = proc.communicate(input='\n'.join(script_data).encode('utf-8'), timeout=30)
+    stdout, _ = proc.communicate(input='\n'.join(script_data).encode('utf-8'), timeout=60)
     if proc.returncode != 0:
         raise ShellcheckRunError(stdout)
 
 
+class Cacher:
+    _instance = None
+    def __new__(cls):
+        if cls._instance is not None:
+            return cls._instance
+        instance = super().__new__(cls)
+        cls._instance = instance
+        instance._enabled = True
+        instance._lock = Lock()
+        instance._hit =0
+        instance._miss = 0
+        return instance
+
+    def disable(self):
+        logging.debug("caching is disabled")
+        self._enabled = False
+
+    @staticmethod
+    def _cache_path_for(digest):
+        prefix = digest[:2]
+        return Path.home().joinpath(".cache", "spread-shellcheck", prefix, digest)
+
+    def is_cached(self, data, path):
+        if not self._enabled:
+            return False, ""
+        h = hashlib.sha256()
+        h.update(data)
+        hdg = binascii.b2a_hex(h.digest()).decode()
+        cachepath = Cacher._cache_path_for(hdg)
+        logging.debug("cache stamp %s, exists? %s", cachepath.as_posix(), cachepath.exists())
+        hit = cachepath.exists()
+        self._record_cache_event(hit)
+        return hit, hdg
+
+    def cache_success(self, digest, path):
+        if not self._enabled:
+            return
+        cachepath = Cacher._cache_path_for(digest)
+        logging.debug("cache success, path %s", cachepath.as_posix())
+        cachepath.parent.mkdir(parents=True, exist_ok=True)
+        cachepath.touch()
+
+    def _record_cache_event(self, hit):
+        with self._lock:
+            if hit:
+                self._hit += 1
+            else:
+                self._miss += 1
+
+    @property
+    def stats(self):
+        return namedtuple('Stats', ['hit', 'miss'])(self._hit, self._miss)
+
+
 def checkfile(path, executor):
     logging.debug("checking file %s", path)
-    with open(path) as inf:
-        data = yaml.safe_load(inf)
+    with open(path, mode='rb') as inf:
+        rawdata = inf.read()
+        cached, digest = Cacher().is_cached(rawdata, path)
+        if cached:
+            logging.debug("entry %s already cached", digest)
```

Comment:
Maybe worth a comment that we only cache positive results so that it's ok to stop here?